src/NServiceBus.RavenDB/Timeouts/TimeoutPersister.cs

```diff
@@ -55,62 +49,59 @@ public TimeoutPersister()
                 results = new List<Tuple<string, DateTime>>();
             }
 
-            RavenQueryStatistics statistics;
+            // default return value for when no results are found
+            nextTimeToRunQuery = DateTime.UtcNow.AddMinutes(10);
 
             using (var session = DocumentStore.OpenSession())
             {
-                int totalCount, skipCount = 0;
+                var query = GetChunkQuery(session)
+                    .Where(t => t.Time > startSlice)
+                    .Select(t => new
+                    {
+                        t.Id,
+                        t.Time
+                    });
 
-                do
+                QueryHeaderInformation qhi;
+                using (var enumerator = session.Advanced.Stream(query, out qhi))
                 {
-                    if (CancellationRequested())
+                    if (qhi.TotalResults < 1000)
                     {
-                        return Enumerable.Empty<Tuple<string, DateTime>>();
+                        logger.DebugFormat("Retrieving {0} timeouts.", qhi.TotalResults);
                     }
-
-                    var query = GetChunkQuery(session);
-
-                    var dueTimeouts =
-                        query.Statistics(out statistics)
-                            .Where(t => t.Time >= startSlice)
-                            .Where(t => t.Time < now)
-                            .Skip(skipCount)
-                            .Select(t => new
-                            {
-                                t.Id,
-                                t.Time
-                            })
-                            .Take(maximumPageSize);
-
-                    foreach (var dueTimeout in dueTimeouts)
+                    else if (qhi.TotalResults < 10000)
                     {
-                        results.Add(new Tuple<string, DateTime>(dueTimeout.Id, dueTimeout.Time));
+                        logger.InfoFormat("Retrieving {0} timeouts. This could take some time.", qhi.TotalResults);
                     }
-
-                    if (CancellationRequested())
+                    else
                     {
-                        return Enumerable.Empty<Tuple<string, DateTime>>();
+                        logger.WarnFormat("Retrieving {0} timeouts. This could take quite some time.", qhi.TotalResults);
                     }
 
-                    skipCount = results.Count + statistics.SkippedResults;
-                    totalCount = statistics.TotalResults;
-                }
-                while (results.Count < totalCount);
+                    while (enumerator.MoveNext())
```

Comment:
Sorry, what I was saying is that NHibernate does an unbounded query. It uses essentially the meta-SQL I posted, and if there are 5 million timeouts waiting in that time range, it will return all of them.

I'm not sure there's any reason to try to be fancier than that for Raven, when any fanciness is likely to have some sort of logical hole in it.